webnn/resources/utils.js

```diff
@@ -1,61 +1,427 @@
+  return array.reduce((accumulator, currentValue) => accumulator * currentValue, 1);
+};
+
+/**
+ * Get JSON resources from specified test resources file.
+ * @param {String} file - A test resources file path
+ * @returns {Object} Test resources
+ */
+const loadResources = (file) => {
+  const loadJSON = () => {
+    let xmlhttp = new XMLHttpRequest();
+    xmlhttp.open("GET", file, false);
+    xmlhttp.overrideMimeType("application/json");
+    xmlhttp.send();
+    if (xmlhttp.status == 200 && xmlhttp.readyState == 4) {
+      return xmlhttp.responseText;
+    } else {
+      throw new Error(`Failed to load ${file}`);
+    }
+  };
+
+  const json = loadJSON();
+  return JSON.parse(json.replace(/\\"|"(?:\\"|[^"])*"|(\/\/.*|\/\*[\s\S]*?\*\/)/g, (m, g) => g ? "" : m));
+};
+
+/**
+ * Get exptected data from given resources with output name.
+ * @param {Array} resources - An array of expected resources
+ * @param {String} outputName - An output name
+ * @returns {Number[]} An expected data array
+ */
+const getExpectedData = (resources, outputName) => {
+  let data;
+  for (let subResources of resources) {
+    if (subResources.name === outputName) {
+      data = subResources.data;
+      break;
+    }
+  }
+  if (data === undefined) {
+    throw new Error(`Failed to get expected data sources by ${outputName}`);
+  }
+  return data;
+};
+
+/**
+ * Get ULP tolerance of conv2d operation.
+ * @param {Object} resources - Resources used for building a graph
+ * @returns {Number} A tolerance number
+ */
+const getConv2dPrecisionTolerance = (resources) => {
+  // number of reduced input elements multiplied by filter and summed (a sliding dot product like pooling)
+  const inputNameArray = Object.keys(resources.inputs);
+  const inputShape = resources.inputs[inputNameArray[0]].shape;
+  const filterShape = resources.inputs[inputNameArray[1]].shape;
+  const options = resources.options;
+  let groups = 1;
+  let inputChannels = inputShape[1]; // default nchw inputLayout
+  let filterWidth = filterShape[3]; // default oihw filterLayout
+  let filterHeight = filterShape[2];
+  if (options) {
+    if (options.groups) {
+      groups = options.groups;
+    }
+    if (options.inputLayout) {
+      if (!['nchw', 'nhwc'].includes(options.inputLayout)) {
+        throw new Error(`Unsupport inputLayout ${options.inputLayout}`);
+      }
+      inputChannels = options.inputLayout === 'nchw' ? inputChannels : inputShape[3];
+    }
+    if (options.filterLayout) {
+      if (!['oihw', 'hwio', 'ohwi', 'ihwo'].includes(options.filterLayout)) {
+        throw new Error(`Unsupport filterLayout ${options.filterLayout}`);
+      }
+      switch (options.filterLayout) {
+        case 'hwio':
+          filterWidth = filterShape[1];
+          filterHeight = filterShape[0];
+          break;
+        case 'ohwi':
+        case 'ihwo':
+          filterWidth = filterShape[2];
+          filterHeight = filterShape[1];
+          break;
+        default:
```

Comment:
I know you already read it above on lines 75 and 76, but it would be good to call it out here for clarity. e.g.

```
case 'oihw':
    // Just use the existing filterWidth and filterHeight above.
    break;
```

Reply:
Added this 'oihw' case in `switch` block, thanks.